Bring 'test_products_found_deployment_report' back to execution

ruda opened the pull request:
Bring middle-ware inspection test back to execution.

Sample output:

```
E       AssertionError: Found 1 unexpected products!                                                                                                                                                                                           
E         Errors are listed below:                                                                                                                                                                                                             
E          Found ['jbossfuse'] but only expected to find                                                                                                                                                                                       
E         ['jbosseap', 'jbosswebserver'] on discovery-mw.
...
FAILED camayoc/tests/qpc/api/v1/reports/test_reports.py::test_products_found_deployment_report[testlab-middleware] - AssertionError: Found 1 unexpected products!
```

Sample configuration:

```yaml
  - name: testlab-middleware
    sources:
      - testlab-middleware
    expected_data:
      discovery-mw:
        distribution:
          release: "Red Hat Enterprise Linux release 9.2"
          name: "Red Hat Enterprise Linux"
          version: "9.2"
        products:
          - name: "JBoss EAP"
            presence: "present"
          - name: "JBoss Web Server"
            presence: "present"
          - name: "JBoss Fuse"
            presence: "absent"
          - name: "JBoss BRMS"
            presence: "absent"

```

Related to JIRA: DISCOVERY-432
https://issues.redhat.com/browse/DISCOVERY-432

ruda committed:
refactor: bring 'test_products_found_deployment_report' back to execution

Related to JIRA: DISCOVERY-432
https://issues.redhat.com/browse/DISCOVERY-432

mirekdlugosz commented:
Just so I understand correctly: this PR brings back middleware test, but in our test lab we have single machine that has middlewares installed AND additional echo in login shell. As a result, a scan running against this machine will report that jbossfuse is installed on machine, which fails the test (because it is not installed and we don't expect to see it). Is that correct?

If that is what is happening, then I'm afraid we can't merge that - it will cause all pipeline runs (nightly and upstream PRs) to fail.

As a immediate workaround we could keep the test skipped. Then locally we can remove skip marker and execute a test, but pipelines are not affected. Of course we should un-skip test in `main` as soon as it can pass on pipelines.

mirekdlugosz commented on camayoc/tests/qpc/api/v1/reports/test_reports.py:
```diff
@@ -116,62 +114,63 @@ def test_products_found_deployment_report(data_provider, scan_info):
-                            entity_info=pformat(entity),
-                        )
+    for hostname, expected_data in scan_info.expected_data.items():
+        expected_products = expected_data.products
+        if not expected_products:
+            continue
+        actual_data = found_hosts.get(hostname)
+        if not actual_data:
+            errors_found.append(
+                f"Host '{hostname}' was expected for scan {scan_info.get('name')}, but not found"
+            )
```
Since we are only supporting Python 3.11, I think we could use walrus operators here (but please check if this is correct):
```suggestion
        if not (expected_products := expected_data.products):
            continue
        if not (actual_data := found_hosts.get(hostname)):
            errors_found.append(
                f"Host '{hostname}' was expected for scan {scan_info.get('name')}, but not found"
            )
```

ruda replied:
Let's use it.

mirekdlugosz reviewed (commented):
Left few suggestions for the code itself.

mirekdlugosz commented on camayoc/tests/qpc/api/v1/reports/test_reports.py:
```diff
@@ -100,10 +99,9 @@ def test_report_content_consistency(data_provider):
 
-@pytest.mark.skip("Skipped until Middleware are handled by Camayoc")
 @pytest.mark.runs_scan
 @pytest.mark.parametrize("scan_info", scan_list(), ids=utils.name_getter)
```
`scan_list()` will return all scans that are of type network, vcenter, satellite. But can we actually detect any middleware on vcenter and satellite?

If not, I would suggest to create new `network_scan_list()` function, that would return network scans only (or maybe `scan_list` could take a param?). This should save some time by not running scans that are irrelevant in this context anyway.

Even if we can detect middlewares on Satellite and VCenter I would consider not running these scans, because our test lab does not have middlewares on them. But that would require a short explanation in comment. Something along the lines of "we can detect middleware on Satellite and VCenter, but we don't have this in our test lab, so we skip them to save some pipeline time".

ruda replied:
Well, I prefered to update `scan_list()` with a custom iterable for filtering source types.

ruda commented:
> Just so I understand correctly: this PR brings back middleware test, but in our test lab we have single machine that has middlewares installed AND additional echo in login shell. As a result, a scan running against this machine will report that jbossfuse is installed on machine, which fails the test (because it is not installed and we don't expect to see it). Is that correct?

Yes, that's correct.

> If that is what is happening, then I'm afraid we can't merge that - it will cause all pipeline runs (nightly and upstream PRs) to fail.

I can disable the echo message at the level of the user that I had set up for login. When I disable the echo message from the bashrc file, then the test passes.

```
camayoc/tests/qpc/api/v1/reports/test_reports.py::test_products_found_deployment_report[testlab-middleware] PASSED [100%]
```

ruda commented on camayoc/tests/qpc/api/v1/reports/test_reports.py:
```diff
@@ -116,62 +114,63 @@ def test_products_found_deployment_report(data_provider, scan_info):
+        for fingerprint in system_fingerprints:
+            present_product_names = []
+            for product in fingerprint.get("products"):
+                name = "".join(product["name"].lower().split())
```
I've being very conservative in this regard (and other code changes). Although, I will adopt your suggestions.

ruda replied:
Yes, much clear.
```
E          Found ['JBoss Fuse'] but only expected to find
E         ['JBoss EAP', 'JBoss Web Server'] on discovery-mw.
```

ruda committed:
feat: scan_list() with customized source types

scan_list can use a custom source type for generating scans.
Also, make use of it with test_products_found_deployment_report.

mirekdlugosz commented:
I am OK with not having coverage for https://issues.redhat.com/browse/DISCOVERY-407 in our test lab. Disabling echo sounds like a good idea to me.

Thanks for including other suggestions, and sorry for late response. I think we can merge this.
